library/core/src/iter/traits/peekable.rs

```diff
@@ -0,0 +1,101 @@
+/// An iterator whose elements may be checked without advancing the iterator.
+///
+/// In general, many [`Iterator`]s are able to "peek" their next element,
+/// showing what would be returned by a call to `next` without advancing their
+/// internal state. If these iterators do not offer such functionality, it can
+/// be manually added to an iterator using the [`peekable`] method.
+///
+/// In most cases, calling [`peekable`] on a [`PeekableIterator`] shouldn't
+/// noticeably affect functionality, however, it's worth pointing out a few
+/// differences between [`peekable`] and [`PeekableIterator`]:
+///
+/// * Stateful iterators like those using [`inspect`](Iterator::inspect) will
+///   eagerly evaluate when peeked by a [`peekable`] wrapper, but may do so
+///   lazily with a custom [`PeekableIterator`] implementation.
+/// * The [`peekable`] wrapper will incur a small performance penalty for
+///   [`next`] and [`next_back`], but [`PeekableIterator`] implementations
+///   incur no such penalty.
+/// * The [`peekable`] wrapper will return a reference to its item, whereas
+///   [`PeekableIterator`] will return the item directly.
+///
+/// Note that this trait is a safe trait and as such does *not* and *cannot*
+/// guarantee that the peeked value will be returned in a subsequent call to
+/// [`next`], no matter how soon the two are called together. A common
+/// example of this is interior mutability; if the interior state of the
+/// iterator is mutated between a call to [`peek`] and [`next`], then the
+/// values may differ.
+///
+/// [`peek`]: Self::peek
+/// [`peekable`]: Iterator::peekable
+/// [`Peekable`]: super::Peekable
+/// [`next`]: Iterator::next
+/// [`next_back`]: DoubleEndedIterator::next_back
+///
+/// # Examples
+///
+/// Basic usage:
+///
+/// ```
+/// #![feature(peekable_iterator)]
+/// use std::iter::PeekableIterator;
+///
+/// // a range knows its current state exactly
+/// let five = 0..5;
+///
+/// assert_eq!(Some(0), five.peek());
+/// ```
+#[unstable(feature = "peekable_iterator", issue = "none")]
+pub trait PeekableIterator: Iterator {
+    /// Returns a reference to the [`next`] value without advancing the iterator.
+    ///
+    /// [`next`]: Iterator::next
+    ///
+    /// # Examples
+    ///
+    /// Basic usage:
+    ///
+    /// ```
+    /// #![feature(peekable_iterator)]
+    /// use std::iter::PeekableIterator;
+    ///
+    /// // a finite range knows its current state exactly
+    /// let five = 0..5;
+    ///
+    /// assert_eq!(Some(0), five.peek());
+    /// ```
+    #[unstable(feature = "peekable_iterator", issue = "none")]
+    fn peek(&self) -> Option<Self::Item>;
```

Comment:
Can you say more about why you picked `Option<Item>` instead of `Option<&Item>` here?  Especially since that's what `peek` does https://doc.rust-lang.org/std/iter/struct.Peekable.html#method.peek

And it'd make sense to me for things like `Repeat` to return a reference to the state, for example.

Reply:
The big reason is that a majority of iterators are completely impossible to do with that type signature. By-reference and by-mutable-reference iterators over collections would completely break, and that's loads of use cases right there.

At first I did that, then I realised how bad of an idea it would be to do that. Essentially, the use cases that would prefer just storing the next value and then referencing it are covered by peekable, and this covers the ones where that seems suboptimal.

Reply:
But that leads to some implementations using `clone()`, and `has_next()` just calls `peek()` which then clones. That calling `has_next()` might allocate is going to be surprising and certainly more costly than `ExactSizeIterator`'s `is_empty()`.